### DISCUSSION TITLE
CPU Hogging

### DISCUSSION THREAD

**Discussion by bashton-Maxar** [nasa/cFS#633] [Q&A]
The change [https://github.com/nasa/HS/pull/38](url) added a call to OS_TaskDelay at the end of the while loop within HS_IdleTask located in hs_custom.c.  When this change was pulled into our project we began seeing notifications of CPU Hogging Detected.  Upon investigation it appears that the introduction of OS_TaskDelay to the IdleTask while loop reduces the 1Hz IdleTaskInterval count to change from ~1.3 billion to ~13,000.  This is executing on a Linux environment with an Intel Xeon Platinum 8259CL CPU @ 2.50GHz.  Given the default values of the parameters used in calculating CurrintUtil in HS_CustomGetUtil (hs_custom.c) this results in an estimated CPU utilization of 9999, even though there is really essentially no utilization.  These parameters can be adjusted to account for this, but we were wondering the following:
1) Has anyone else observed this behavior?
2) Is there any concern with the apparent overhead ramifications of using OS_TaskDelay(0) on the IdleTask?
3) Is there any plan to change how the IdleTask operates?

**Comment by skliper** [comment]
> Is there any plan to change how the IdleTask operates?

Yes please.  Related issues: https://github.com/nasa/HS/issues/3, https://github.com/nasa/HS/issues/4.

Suggested approach is to move CPU idle API to the PSP and use appropriate functions based on the platform to get the idle stats.  Or if there is standard OS support move it to OSAL.  Really doesn't need to be in HS, and abstracting it will allow for more appropriate implementation based on the environment.
